lib/optimizely/optimizely_user_context.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+#
+#    Copyright 2020, Optimizely and contributors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'json'
+
+module Optimizely
+  class OptimizelyUserContext
+    # Representation of an Optimizely User Context using which APIs are to be called.
+
+    attr_reader :user_id, :user_attributes
+
+    def initialize(optimizely_client, user_id, user_attributes)
+      @optimizely_client = optimizely_client
+      @user_id = user_id
+      @user_attributes = user_attributes.nil? ? {} : user_attributes.clone
+    end
+
+    def set_attribute(attribute_key, attribute_value)
+      @user_attributes[attribute_key] = attribute_value
+    end
+
+    def decide(key, options = nil)
+      @optimizely_client&.decide(self, key, options)
+    end
+
+    def decide_for_keys(keys, options = nil)
+      @optimizely_client&.decide_for_keys(self, keys, options)
+    end
+
+    def decide_all(options = nil)
+      @optimizely_client&.decide_all(self, options)
```

Comment:
don't you want to check if optmizely_client is nil and log a message or pass an error?

Reply:
Good Point! I thought about it and i believe that ruby being a dynamic language, this is possible if someone is determined to produce this scenario but this is not possible in the intended usage. The only way we expect users to obtain user_context instance is by calling a `get_user_context` method on the optimizely instance itself, which means the optimizely_instance cant be null. I believe it would be overkill to add a check and a log message.

If we still want to do that, i am a little worried about where would i get logger from. I will probably need to pass in the logger in the user_context constructor along with the optimizely_client which is doable but i really dont think its necessary because if a user has generated user_context through a method on `optimizely` instance, it cant simply be null.

What do you think?